File: test/python/circuit/library/test_quantum_volume.py
# This code is part of Qiskit.
#
# (C) Copyright IBM 2017, 2020.
#
# This code is licensed under the Apache License, Version 2.0. You may
# obtain a copy of this license in the LICENSE.txt file in the root directory
# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
#
# Any modifications or derivative works of this code must retain this
# copyright notice, and modified files need to carry a notice indicating
# that they have been altered from the originals.

"""Test library of quantum volume circuits."""

import unittest

from test.utils.base import QiskitTestCase
from qiskit.circuit.library import QuantumVolume
from qiskit.circuit.library.quantum_volume import quantum_volume


class TestQuantumVolumeLibrary(QiskitTestCase):
    """Test library of quantum volume quantum circuits."""

    def test_qv_seed_reproducibility(self):
        """Test qv circuit."""
        left = QuantumVolume(4, 4, seed=28, classical_permutation=False)
        right = QuantumVolume(4, 4, seed=28, classical_permutation=False)
        self.assertEqual(left, right)

        left = QuantumVolume(4, 4, seed=3, classical_permutation=True)
        right = QuantumVolume(4, 4, seed=3, classical_permutation=True)
        self.assertEqual(left, right)

        left = QuantumVolume(4, 4, seed=2024, flatten=True)
        right = QuantumVolume(4, 4, seed=2024, flatten=True)
        self.assertEqual(left, right)

    def test_qv_function_seed_reproducibility(self):
        """Test qv circuit."""
        left = quantum_volume(10, 10, seed=128)
        right = quantum_volume(10, 10, seed=128)
        self.assertEqual(left, right)

        left = quantum_volume(10, 10, seed=256)
        right = quantum_volume(10, 10, seed=256)
        self.assertEqual(left, right)

        left = quantum_volume(10, 10, seed=4196)
        right = quantum_volume(10, 10, seed=4196)
        self.assertEqual(left, right)


if __name__ == "__main__":
    unittest.main()

File: test/c/test_quantum_volume.c
// This code is part of Qiskit.
//
// (C) Copyright IBM 2025.
//
// This code is licensed under the Apache License, Version 2.0. You may
// obtain a copy of this license in the LICENSE.txt file in the root directory
// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
//
// Any modifications or derivative works of this code must retain this
// copyright notice, and modified files need to carry a notice indicating
// that they have been altered from the originals.

#include "common.h"
#include <complex.h>
#include <math.h>
#include <qiskit.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <string.h>

/**
 * Test building a quantum volume circuit
 */
int test_qv(void) {
    const uint32_t num_qubits = 5;
    int result = Ok;
    QkCircuit *qv = qk_circuit_library_quantum_volume(num_qubits, num_qubits, 42);
    size_t num_instructions = qk_circuit_num_instructions(qv);
    if (num_instructions != 10) {
        printf("Unexpected number of instructions: %zu", num_instructions);
        result = EqualityError;
        goto cleanup;
    }
    QkCircuitInstruction inst;
    for (size_t i = 0; i < num_instructions; i++) {
        qk_circuit_get_instruction(qv, i, &inst);
        int cmp_result = strcmp(inst.name, "unitary");
        if (cmp_result != 0) {
            result = EqualityError;
            goto loop_exit;
        }
        if (inst.num_qubits != 2) {
            result = EqualityError;
            goto loop_exit;
        }
    loop_exit:
        qk_circuit_instruction_clear(&inst);
        if (result != 0) {
            break;
        }
    }

cleanup:
    qk_circuit_free(qv);
    return result;
}

int test_quantum_volume(void) {
    int num_failed = 0;
    num_failed += RUN_TEST(test_qv);

    fflush(stderr);
    fprintf(stderr, "=== Number of failed subtests: %i\n", num_failed);

    return num_failed;
}
